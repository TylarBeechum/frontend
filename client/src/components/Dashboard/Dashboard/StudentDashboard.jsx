import React from 'react';
import { useStateValue } from 'react-conflux';
import styled from 'styled-components';
import { InfiniteScroll, Box } from 'grommet';
import FuzzySearch from 'fuzzy-search';

import searchIcon from '../../../images/search-icon.svg';
import CredCard from '../Card/CredCard';
import DashboardLoading from '../DashboardLoading';

import {
  studentContext,
  SEARCH_HANDLE_CHANGE
} from '../../../store/reducers/studentReducer';

const StudentDashboard = () => {
  const [studentState, studentDispatch] = useStateValue(studentContext);
  let searchResult = [];
  if (studentState.studentData) {
    const searchTerms = ['credName', 'criteria', 'ownerName', 'issuedOn'];
    const searchOptions = {
      caseSensitive: false
    };
    const searcher = new FuzzySearch(
      studentState.studentData.studentDetails.credentials,
      searchTerms,
      searchOptions
    );
    searchResult = searcher.search(studentState.studentSearchInput);
  }
  const handleChange = e => {
    studentDispatch({ type: SEARCH_HANDLE_CHANGE, payload: e.target.value });
  };
  if (!studentState.studentData) return <div />;
  return (
    <>
      <StudentDetails>
        {studentState.studentDataSuccess ? (
          <h2>Welcome, {studentState.studentData.studentDetails.fullName}!</h2>
        ) : (
          <div />
        )}
        <div>
          <input
            type="text"
            name="searchText"
            placeholder="Search"
            onChange={handleChange}
            value={studentState.studentSearchInput}
          />
        </div>
      </StudentDetails>
      {studentState.studentDataStart ? (
        <DashboardLoading />
      ) : (
        <>
          {studentState.studentDataSuccess && searchResult.length ? (
            <CredCardContainer>
              <InfiniteScroll items={searchResult} step={10}>
                {item => {
                  return <CredCard key={item.id} cred={item} />;
                }}
              </InfiniteScroll>
            </CredCardContainer>
          ) : (
            studentState.studentDataSuccess && (
              <NothingFound>
                {!studentState.studentData.studentDetails.credentials.length
                  ? "You don't have any credentials yet..."
                  : 'No results were found..'}
              </NothingFound>
            )
          )}
        </>
      )}
    </>
  );
};

const StudentDetails = styled.section`
  margin: 50px auto 30px;
  padding: 0 2%;
  max-width: 1675px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;

  div {
    display: flex;
    width: 68%;
    justify-content: flex-end;
    align-items: center;

    input {
      background: white url(${searchIcon}) no-repeat scroll 5px 5px;
      background-size: 20px;
      background-position: right 13px center;
      border: 2px solid ${({ theme }) => theme.global.colors.searchBarBorder};
      color: ${({ theme }) => theme.global.colors.searchBarColor};
      height: 40px;
      margin-right: 1.5%;
      padding: 25px;
      transition: 0.25s ease-in-out;
      width: 220px;
      caret-color: grey;
      border-radius: 50px;
      font-size: 1.8rem;

      ::placeholder {
        color: ${({ theme }) => theme.global.colors.searchBarColor};
        font-weight: 700;
      }

      &:focus {
        color: ${({ theme }) => theme.global.colors.brand};
        border-color: ${({ theme }) => theme.global.colors.brand};
        outline: none;
        padding-left: 15px;
        width: 50%;
      }
    }
  }
`;

const NothingFound = styled.p`
  width: 100%;
  text-align: center;
  font-size: 2.4rem;
  margin-top: 20vh;
  color: ${({ theme }) => theme.global.colors.dashboardNotFoundColor};
`;

const CredCardContainer = styled(Box)`
  height: 75vh;
  overflow: auto;
  padding: 10px 2%;
`;


export default StudentDashboard;
