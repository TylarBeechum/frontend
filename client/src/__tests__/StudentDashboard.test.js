/* eslint-disable */
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {
  renderWithRouterAndProviders,
  rendererWithRouterAndProviders
} from '../__testHelpers__';

import StudentDashboard from '../components/Dashboard/Dashboard/StudentDashboard';

describe('<StudentDashboard />', () => {
  describe('should render the component successfully without crashing', () => {
    it('tests that the component does not crash when it is rendered', () => {
      const helpers = renderWithRouterAndProviders(
        <StudentDashboard history={{ location: { pathname: '/' } }} />
      );
    });
  });

  describe('should match the previous snapshot of the component', () => {
    it('tests that the component has not changed and matched the previous snapshot', () => {
      const tree = rendererWithRouterAndProviders(
        <StudentDashboard history={{ location: { pathname: '/' } }} />
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('should render the credentials for each school to the screen', () => {
    it("tests that the component renders the school's name above the credentials", () => {
      const { getByText } = renderWithRouterAndProviders(
        <StudentDashboard history={{ location: { pathname: '/' } }} />
      );
      const cred1 = getByText(/test1/i);
      const cred2 = getByText(/test2/i);
      const cred3 = getByText(/test3/i);
      expect(cred1).toBeDefined();
      expect(cred2).toBeDefined();
      expect(cred3).toBeDefined();
    });
  });

  describe("should render all credentials' view buttons to the screen", () => {
    it('tests that all credentials have their associated View buttons', () => {
      const { getAllByText } = renderWithRouterAndProviders(
        <StudentDashboard history={{ location: { pathname: '/' } }} />
      );
      const viewBtns = getAllByText(/View/i);
      expect(viewBtns).toHaveLength(3);
    });
  });
});
