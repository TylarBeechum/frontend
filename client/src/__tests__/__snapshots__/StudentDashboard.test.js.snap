// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudentDashboard /> should match the previous snapshot of the component tests that the component has not changed and matched the previous snapshot 1`] = `
<div
  className="StyledGrommet-sc-19lkkz7-0 iDwiuh"
>
  <section
    className="sc-cSHVUG chVKOk"
  >
    <h2>
      Welcome, 
      !
    </h2>
    <div>
      <input
        name="searchText"
        onChange={[Function]}
        placeholder="Search"
        type="text"
        value=""
      />
    </div>
  </section>
  <div
    className="StyledBox-sc-13pk1d4-0 izmUvx"
  >
    <section
      className="sc-VigVT kwgGCs"
    >
      <div
        className="sc-EHOje hWneJQ"
      >
        <button
          aria-label="view credentials button"
          className="sc-bdVaJa sc-bZQynM csIfUH StyledButton-sc-323bzc-0 kHrfdt"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          View
        </button>
      </div>
      <div
        className="sc-gzVnrw bKsZFP"
      >
        <h2>
          test1
        </h2>
        <p>
          test
        </p>
      </div>
      <div
        className="sc-htoDjs erhmUl"
      >
        <p />
      </div>
      <div
        className="sc-dnqmqq jubHKb"
      >
        <h3>
          Nathan Thomas
        </h3>
      </div>
      <div
        className="sc-iwsKbI eNHtrI"
      >
        <button
          className="sc-bdVaJa sc-gqjmRU hnZXkc StyledButton-sc-323bzc-0 jfbKsv"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          <svg
            aria-label="Trash"
            className="sc-gZMcBi Qfbta StyledIcon-ofa7kd-0 iOkQrb"
            viewBox="0 0 24 24"
          >
            <path
              d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
    </section>
    <section
      className="sc-VigVT kwgGCs"
    >
      <div
        className="sc-EHOje hWneJQ"
      >
        <button
          aria-label="view credentials button"
          className="sc-bdVaJa sc-bZQynM csIfUH StyledButton-sc-323bzc-0 kHrfdt"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          View
        </button>
      </div>
      <div
        className="sc-gzVnrw bKsZFP"
      >
        <h2>
          test2
        </h2>
        <p>
          test
        </p>
      </div>
      <div
        className="sc-htoDjs erhmUl"
      >
        <p />
      </div>
      <div
        className="sc-dnqmqq jubHKb"
      >
        <h3>
          Nathan Thomas
        </h3>
      </div>
      <div
        className="sc-iwsKbI eNHtrI"
      >
        <button
          className="sc-bdVaJa sc-gqjmRU hnZXkc StyledButton-sc-323bzc-0 jfbKsv"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          <svg
            aria-label="Trash"
            className="sc-gZMcBi Qfbta StyledIcon-ofa7kd-0 iOkQrb"
            viewBox="0 0 24 24"
          >
            <path
              d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
    </section>
    <section
      className="sc-VigVT kwgGCs"
    >
      <div
        className="sc-EHOje hWneJQ"
      >
        <button
          aria-label="view credentials button"
          className="sc-bdVaJa sc-bZQynM csIfUH StyledButton-sc-323bzc-0 kHrfdt"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          View
        </button>
      </div>
      <div
        className="sc-gzVnrw bKsZFP"
      >
        <h2>
          test3
        </h2>
        <p>
          test
        </p>
      </div>
      <div
        className="sc-htoDjs erhmUl"
      >
        <p />
      </div>
      <div
        className="sc-dnqmqq jubHKb"
      >
        <h3>
          Nathan Thomas
        </h3>
      </div>
      <div
        className="sc-iwsKbI eNHtrI"
      >
        <button
          className="sc-bdVaJa sc-gqjmRU hnZXkc StyledButton-sc-323bzc-0 jfbKsv"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          <svg
            aria-label="Trash"
            className="sc-gZMcBi Qfbta StyledIcon-ofa7kd-0 iOkQrb"
            viewBox="0 0 24 24"
          >
            <path
              d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
    </section>
  </div>
</div>
`;
